src/backend/PortabilityService.Functions/Report.cs

```diff
@@ -3,100 +3,115 @@
 
 using Microsoft.Azure.WebJobs;
 using Microsoft.Azure.WebJobs.Extensions.Http;
-using Microsoft.Azure.WebJobs.Host;
+using Microsoft.Extensions.Logging;
+using Microsoft.Fx.Portability;
+using Microsoft.Fx.Portability.Azure;
+using Microsoft.Fx.Portability.ObjectModel;
+using Microsoft.Fx.Portability.Reporting;
+using Microsoft.Fx.Portability.Reports;
+using Microsoft.WindowsAzure.Storage;
+using PortabilityService.Functions.DependencyInjection;
 using System;
+using System.Collections.Generic;
 using System.IO;
 using System.Linq;
 using System.Net;
 using System.Net.Http;
 using System.Net.Http.Headers;
+using System.Threading.Tasks;
 
 namespace PortabilityService.Functions
 {
     public static class Report
     {
         [FunctionName("report")]
-        public static HttpResponseMessage Run(
+        public static async Task<HttpResponseMessage> Run(
             [HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = "report/{submissionId}")] HttpRequestMessage req,
             string submissionId,
-            TraceWriter log)
+            [Inject] IStorage storage,
+            [Inject] IReportTokenValidator validator,
+            ILogger log)
         {
-            if (!ValidAccessKey(req))
+            if (!validator.RequestHasValidToken(req))
             {
                 return req.CreateResponse(HttpStatusCode.Unauthorized);
             }
 
-            // simulate report generation taking some time
-            if (new Random().Next(10) < 4)
+            var (analyzeRequest, analyzeResult) = await GetSubmissionDataAsync(submissionId, storage, log);
+            if (analyzeRequest == null)
             {
-                // TODO this should return 202 only when the analyze request is known to have been received,
-                // 404 otherwise (e.g. when the client tries to get the report before its analyze request
-                // has propagated)
-                var response = req.CreateResponse(HttpStatusCode.Accepted);
-                response.Headers.RetryAfter = new RetryConditionHeaderValue(TimeSpan.FromSeconds(2d));
+                // the request may have been received but not yet stored => the client should poll again
+                var res = req.CreateResponse(HttpStatusCode.NotFound);
+                res.Headers.RetryAfter = new RetryConditionHeaderValue(TimeSpan.FromSeconds(4d));
```

Comment:
why the `d`?

Reply:
I know it's unnecessary, I somehow developed the habit of being pedantic about literals like this.